File: Week2/hw2_var2.cpp
#include <iostream>
#include <cmath>
#include <chrono>
using namespace std::chrono;


// 1D Heat equation
// Thomas algorithm
// Variant 2

// Initial condition
// U(t=0, x) = 1 - x^3

// Boundary conditions
// Ux(t, x=0) = 0
// U(t, x=1) = 0

const double pi = 3.14159265359;
const double dx = 0.01;
const double dt = 0.01;
const double start_x = 0, end_x = 1;
const int N = (end_x - start_x) / dx + 1;

const double eps = 1e-6;
const int stop_iteration = 3e5;

const double A = -1.0 / (dx*dx);
const double B = 1.0 / dt + 2.0 / (dx*dx);
const double C = -1.0 / (dx*dx);

double x[N], F[N], D[N], alpha[N], beta[N];
double U_old[N], U_new[N];

void Analytical_solution(double t, int m=100){
    for(int n = 1; n <= m; ++n){
        for(int i = 0; i < N; ++i){
            F[i] += -96*(pow(-1, n) + 2/((2*n - 1)*pi))
                / pow((2*n - 1)*pi, 3)
                * exp(-pow((2*n - 1)*pi/2, 2)*t)
                * cos((2*n - 1)*pi/2*x[i]);
        }
    }    
}

int main(){
    
    for(int i = 0; i < N; ++i){
        x[i] = start_x + i*dx;
    }

    // Initialize arrays
    for(int i = 0; i < N; ++i){
        F[i] = 0;
        U_old[i] = 0;
        U_new[i] = 0;
        alpha[i] = 0;
        beta[i] = 0;
    }

    // initial condition
    // U(x, t=0) = 1 - x^3
    for(int i = 0; i < N; ++i){
        U_old[i] = 1 - pow(x[i], 3);
    }

    // Finding D
    for(int i = 0; i < N; ++i){
        D[i] = U_old[i] / dt;
    }

    auto start = high_resolution_clock::now();

    int iteration = 0;
    double maximum = 0;
    do{
        for(int i = 0; i < N; ++i){
            D[i] = U_old[i] / dt;
        }

        alpha[1] = 1;
        beta[1] = 0;
        for(int i = 1; i < N-1; ++i){
            alpha[i+1] = -A / (B + C*alpha[i]);
            beta[i+1] = (D[i] - C*beta[i]) / (B + C*alpha[i]);
        }

        U_new[N-1] = 0;
        for(int i = N-2; i >= 0; i--){
            U_new[i] = alpha[i+1]*U_new[i+1] + beta[i+1];
        }

        maximum = 0;
        for(int i = 0; i < N; ++i){
            if(maximum < fabs(U_new[i] - U_old[i])){
                maximum = fabs(U_new[i] - U_old[i]);
            }
        }

        for(int i = 0; i < N; ++i){
            U_old[i] = U_new[i];
        }

        iteration++;
    }while(maximum > eps and iteration < stop_iteration);

    // Finding analytical solution
    Analytical_solution(iteration * dt);

    maximum = 0;
    for(int i = 0; i < N; ++i){
        if(maximum < fabs(F[i] - U_new[i])){
            maximum = fabs(F[i] - U_new[i]);
        }
    }

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    
    printf("calculating time: %.6f seconds\n", duration.count() / 1e6);
    printf("Maximum difference: %.9f\n", maximum);
    printf("Number of iterations: %d\n", iteration);

    FILE *File;
    File = freopen("HW2_X_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.9f", x[i]);
        }else{
            printf("%.9f\t", x[i]);
        }
    }
    fclose(File);

    File = freopen("HW2_U_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.6e", U_new[i]);
        }else{
            printf("%.6e\t", U_new[i]);
        }
    }
    fclose(File);

    File = freopen("HW2_F_cpp.txt", "w", stdout);
    for(int i = 0; i < N; ++i){
        if(i == N-1){
            printf("%.6e", F[i]);
        }else{
            printf("%.6e\t", F[i]);
        }
    }
    fclose(File);

    return 0;
}

File: Week2/hw2_var2.py
import numpy as np
import matplotlib.pyplot as plt
from numpy import (exp, pi, sin, cos)
from time import perf_counter


# 1D Heat equation
# Thomas algorithm
# Variant 2

# Initial condition
# U(t=0, x) = 1 - x^3

# Boundary conditions
# Ux(t, x=0) = 0
# U(t, x=1) = 0

def Analytical_solution(x, t, m=100):
    λ = lambda n: (2*n - 1)*pi / 2
    A_n = lambda n: -12*((-1)**n + 1/λ(n)) / λ(n)**3
    
    F = 0
    for n in range(1, m+1):
        F += A_n(n) * exp(-λ(n)**2*t) * cos(λ(n)*x)
    
    return F

dx = 0.01
dt = 0.01
start_x, end_x = (0, 1)
N = int((end_x - start_x) / dx) + 1
x = start_x + np.arange(start=0, stop=N) * dx

eps = 1e-9
stop_iteration = 5e4

alpha = np.zeros_like(x)
beta = np.zeros_like(x)
U_old = np.zeros_like(x)
U_new = np.zeros_like(x)

# Initial condition
U_old[:] = 1 - x**3

A = -1 / dx**2
B = 1 / dt + 2 / dx**2
C = -1 / dx**2
D = U_old / dt # numpy array

start_time = perf_counter()

iteration = 0
maximum = 1
while maximum > eps and iteration < stop_iteration:
    D[:] = U_old / dt # numpy array
    
    alpha[1] = 1
    beta[1] = 0
    for i in range(1, N-1):
        alpha[i+1] = -A / (B + C*alpha[i])
        beta[i+1] = (D[i] - C*beta[i]) / (B + C*alpha[i])
    
    U_new[N-1] = 0
    for i in range(N-2, -1, -1):
        U_new[i] = alpha[i+1]*U_new[i+1] + beta[i+1]
    
    maximum = np.max(np.abs(U_new - U_old))
    
    U_old = U_new.copy()
    iteration += 1

F = Analytical_solution(x, dt*iteration)
end_time = perf_counter()

print(f"Calculation time: {end_time - start_time:.6f} seconds")
print(f"Maximum difference: {np.max(np.abs(F - U_new)):.9f}")
print(f"Number of iterations: {iteration}")

np.savetxt("HW2_X_py.txt", x, fmt="%.6e", delimiter="\t")
np.savetxt("HW2_U_py.txt", U_new, fmt="%.6e", delimiter="\t")
np.savetxt("HW2_F_py.txt", F, fmt="%.6e", delimiter="\t")

plt.title("Heat equation using tridiagonal matrix method")
plt.grid()
plt.plot(x, U_new, label="Numerical solution")
plt.plot(x, F, ls="--", label="Analytical solution")
plt.xlabel("x")
plt.ylabel("U(x)")
plt.legend()
plt.tight_layout()
plt.show()
